tools/docker/Dockerfile

```diff
@@ -0,0 +1,184 @@
+ENV PATH /bin:/usr/bin:/usr/local/bin:/sbin:/usr/sbin:/usr/local/sbin:/usr/lib/go/bin:/root/.cargo/bin
+RUN --mount=type=cache,id=aptcache,target=/var/cache/apt --mount=type=cache,id=aptlib,target=/var/lib/apt \
+mkdir -p /etc/apt/sources.list.d \
+&& echo "deb http://ftp.debian.org/debian stretch-backports main" > /etc/apt/sources.list.d/stretch-backports.list \
+&& apt update \
+&& apt --target-release stretch-backports install -y golang-go \
+&& apt install -y \
+git \
+curl \
+build-essential \
+libseccomp-dev \
+btrfs-tools \
+libdevmapper-dev \
+musl-tools \
+util-linux \
+ca-certificates
+
+
+
+
+# Build firecracker itself
+# TODO can we cache cargo artifacts like we do go modules?
+FROM build-base as firecracker-build
+SHELL ["/bin/bash", "-c"]
+RUN curl https://sh.rustup.rs -sSf | sh -s -- -y --default-toolchain 1.32.0 \
+&& source /root/.cargo/env \
+&& rustup target add x86_64-unknown-linux-musl \
+&& git clone https://github.com/firecracker-microvm/firecracker.git \
+&& cd firecracker \
+&& git checkout v0.15.2 \
+&& cargo build --release --features vsock --target x86_64-unknown-linux-musl \
+&& mkdir /output \
+&& cp target/x86_64-unknown-linux-musl/release/firecracker /output \
+&& cp target/x86_64-unknown-linux-musl/release/jailer /output
+
+
+
+
+# Build containerd
+FROM build-base as containerd-build
+RUN --mount=id=gomod,target=/root/go/pkg/mod,type=cache mkdir -p /root/go/src/github.com/containerd/containerd \
+&& git clone https://github.com/containerd/containerd.git /root/go/src/github.com/containerd/containerd \
+&& cd /root/go/src/github.com/containerd/containerd \
+&& git checkout v1.2.4 \
+&& make \
+&& mkdir /output && cp -R /root/go/src/github.com/containerd/containerd/bin/* /output
+
+
+
+
+# Build runc
+FROM build-base as runc-build
+RUN --mount=id=gomod,target=/root/go/pkg/mod,type=cache mkdir -p /root/go/src/github.com/opencontainers/runc \
+&& git clone https://github.com/opencontainers/runc /root/go/src/github.com/opencontainers/runc \
+&& cd /root/go/src/github.com/opencontainers/runc \
+&& git checkout 6635b4f0c6af3810594d2770f662f34ddc15b40d \
+&& make static BUILDTAGS='seccomp' \
+&& mkdir /output && make BINDIR='/output' install
+
+
+
+
+# Build firecracker-containerd
+FROM build-base as firecracker-containerd-build
+ADD . /firecracker-containerd
+RUN --mount=id=gomod,target=/root/go/pkg/mod,type=cache cd /firecracker-containerd \
+&& make STATIC_AGENT='true' \
+&& mkdir -p /output \
+&& cp agent/agent runtime/containerd-shim-aws-firecracker snapshotter/cmd/devmapper/devmapper_snapshotter snapshotter/cmd/naive/naive_snapshotter /output
+
+
+
+
+# Build a rootfs for the microVM, including runc and firecracker-containerd's agent
+FROM alpine:3.8 as firecracker-vm-root
+COPY --from=runc-build /output/runc /usr/local/bin/
+COPY --from=firecracker-containerd-build /output/agent /usr/local/bin/
+ADD tools/docker/fc-agent.start /etc/local.d/fc-agent.start
+RUN apk add musl \
+busybox \
+alpine-baselayout \
```

Comment:
Are all of these needed? Like why is ncurses needed?

Reply:
This list was obtained by just getting the list of packages installed by firecracker's "hello-world" rootfs and then removing some SSH packages which I knew for sure were not needed.

I'll try out removing some more like ncurses to get it down to the bare minimum